cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java

```diff
@@ -781,4 +790,102 @@ private void writeSystemMetadata(Id.Artifact artifactId, ArtifactInfo artifactIn
+                                                          Id.Artifact artifactId, String pluginType,
+                                                          String pluginName, String methodName,
+                                                          String parameters) throws ArtifactNotFoundException,
+    IOException, IllegalArgumentException, BadRequestException, ClassNotFoundException, InvocationTargetException {
+    ArtifactDetail artifactDetail = artifactStore.getArtifact(artifactId);
+    return invokePluginEndpoint(namespace, artifactDetail, pluginType, pluginName,
+                                pickParentArtifact(artifactDetail), methodName, parameters);
+  }
+
+  private ArtifactDescriptor pickParentArtifact(ArtifactDetail artifactDetail)
+    throws InvalidArtifactException, ArtifactNotFoundException, IOException {
+
+    // get parent artifacts
+    Set<ArtifactRange> parentArtifactRanges = artifactDetail.getMeta().getUsableBy();
+    if (parentArtifactRanges.isEmpty()) {
+      throw new IllegalArgumentException("No parent artifact found for the plugin artifact");
+    }
+
+    // just pick the first parent artifact from the set.
+    ArtifactRange parentArtifactRange = parentArtifactRanges.iterator().next();
+    List<ArtifactDetail> artifactDetails = artifactStore.getArtifacts(parentArtifactRange);
+    if (artifactDetails.isEmpty()) {
+      // should not happen
+      throw new InvalidArtifactException(String.format("No Artifact found in the range %s", parentArtifactRange));
+    }
+
+    // return the first one from the artifact details list.
+    return artifactDetails.get(0).getDescriptor();
+  }
+
+  private String invokePluginEndpoint(Id.Namespace namespace, ArtifactDetail artifactDetail, String pluginType,
+                                      String pluginName, ArtifactDescriptor parentArtifactDescriptor,
+                                      String methodName, String requestBody)
+    throws NotFoundException, IOException, BadRequestException, ClassNotFoundException, InvocationTargetException {
+
+    try (CloseableClassLoader parentClassLoader =
+           artifactClassLoaderFactory.createClassLoader(parentArtifactDescriptor.getLocation())) {
+      Id.Artifact artifactId = Id.Artifact.from(namespace, artifactDetail.getDescriptor().getArtifactId());
+      File stageDir = DirUtils.createTempDir(baseUnpackDir);
+
+      try (PluginInstantiator pluginInstantiator = new PluginInstantiator(cConf, parentClassLoader, stageDir)) {
+
+        Set<PluginClass> pluginClasses = artifactDetail.getMeta().getClasses().getPlugins();
+
+        for (PluginClass pluginClass : pluginClasses) {
+          if (!pluginClass.getName().equals(pluginName) || !(pluginClass.getType().equals(pluginType))) {
+            continue;
+          }
+          // plugin type and name matched, next check for endpoint method presence
+          if (pluginClass.getEndpoints() == null || !pluginClass.getEndpoints().contains(methodName)) {
```

Comment:
finding and verifiying existence should also happen before you do any file operations like creating a temp dir and creating an PluginInstantiator